### PR TITLE
simplify pagination API

### PR DESCRIPTION
**Pull request by gavinking**
Implements the solution suggested by @njr-11

It seems to me that this design reduces:

1. understandability from the point of view of the user
2. burden on implementors
3. burden on development and maintenance of the TCK
4. constraints on future evolution of the API

This for a rather small loss of typesafety.

See #504.

### CODE REVIEW

**Comment by gavinking** [comment]
I have now also pulled the operations for reverse-navigation up to `Page`, since there's no reason why offset-based pagination has to go only in a forward direction.

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:79]:
```suggestion
     * If the number of elements in the page is less than the maximum page size,
     * then there are no subsequent pages of data to read.
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:103]:
```suggestion
     * page was obtained.
```

File: api/src/main/java/jakarta/data/page/PageRequest.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/PageRequest.java:453]: The JavaDoc for the other methods states IllegalStateException.
```suggestion
     * throw an {@link IllegalStateException} when called.
```

File: api/src/main/java/jakarta/data/page/PageRequest.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/PageRequest.java:464]:
```suggestion
     *         set to {@code true}.
```

File: api/src/main/java/jakarta/data/page/package-info.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/package-info.java:50]:
```suggestion
 * <p>The elements of a {@link jakarta.data.page.Page} are computed based on their positional offset
```

File: tck/src/main/java/ee/jakarta/tck/data/framework/read/only/AsciiCharacters.java
Comment (njr-11) [inline tck/src/main/java/ee/jakarta/tck/data/framework/read/only/AsciiCharacters.java:67]: Thank you for commenting this out rather than removing it and the related tests. In a follow-on pull, I should be able to convert this into a method that returns List and then obtain multiple streams from that in order to make the tests valid again.

File: tck/src/main/java/ee/jakarta/tck/data/framework/read/only/PositiveIntegers.java
Comment (njr-11) [inline tck/src/main/java/ee/jakarta/tck/data/framework/read/only/PositiveIntegers.java:50]: A return type of Page would require a PageRequest parameter.
```suggestion
    List<NaturalNumber> findByIsOddFalseAndIdBetween(long min, long max);
```

File: tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java
Comment (njr-11) [inline tck/src/main/java/ee/jakarta/tck/data/standalone/entity/EntityTests.java:536]: Cannot be Page because it has no PageRequest parameter
```suggestion
        List<NaturalNumber> even = positives.findByIsOddFalseAndIdBetween(50L, 60L);
```

File: tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/Catalog.java
Comment (njr-11) [inline tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/Catalog.java:76]: Cannot be Page because it has no PageRequest
```suggestion
    List<Product> findByDepartmentCountAndPriceBelow(int numDepartments, double maxPrice);
```

File: tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/PersistenceEntityTests.java
Comment (njr-11) [inline tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/PersistenceEntityTests.java:317]: Cannot be page because it has no PageRequest
```suggestion
        List<Product> found = catalog.findByDepartmentCountAndPriceBelow(2, 100.0);
```

**Review by njr-11 — COMMENTED** [review]
This looks great!  I spotted a few minor typos and a few remaining places where slice can be switched to page, and a couple of places where Streamable was switched to Page but needs to instead be switched to List because the repository method has no PageRequest parameter.

File: api/src/main/java/jakarta/data/page/PageRequest.java
Comment (gavinking) [reply api/src/main/java/jakarta/data/page/PageRequest.java:453]: Oops 🫣that is definitely what I meant.

File: tck/src/main/java/ee/jakarta/tck/data/framework/read/only/PositiveIntegers.java
Comment (gavinking) [reply tck/src/main/java/ee/jakarta/tck/data/framework/read/only/PositiveIntegers.java:50]: Yeah sorry, good point.

**Comment by gavinking** [comment]
@njr-11 I applied all your suggestions.

Now I will try to resolve the conflicts.

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:127]:
```suggestion
     * Returns a request for the {@linkplain PageRequest#next() next} page if
     * {@link #hasNext()} indicates there might be a next page.
     *
     * @return a request for the next page.
     * @throws NoSuchElementException if it is known that there is no next page.
     *         To avoid this exception, check for a {@code true} result of
     *         {@link #hasNext()} before invoking this method.
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:133]:
```suggestion
     * <p>Returns a request for the {@linkplain PageRequest#next() next} page if
     * {@link #hasNext()} indicates there might be a next page.</p>
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:142]:
```suggestion
     * @return a request for the next page.
     * @throws NoSuchElementException if it is known that there is no next page.
     *         To avoid this exception, check for a {@code true} result of
     *         {@link #hasNext()} before invoking this method.
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:150]:
```suggestion
     * <p>Returns a request for the {@link PageRequest#previous() previous} page,
     * if {@link #hasPrevious()} indicates there might be a previous page.</p>
     *
     * @return a request for the previous page.
     * @throws NoSuchElementException if it is known that there is no previous page.
     *         To avoid this exception, check for a {@code true} result of
     *         {@link #hasPrevious()} before invoking this method.
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (njr-11) [inline api/src/main/java/jakarta/data/page/Page.java:165]:
```suggestion
     * <p>Returns a request for the {@linkplain PageRequest#previous() previous}
     * page, if {@link #hasPrevious()} indicates there might be a previous page.</p>
     *
     * <p>This method is useful when {@linkplain Query query language} is used to
     * return a result of different type than the entity that is being queried.
     * This method allows the subsequent {@link PageRequest} to be returned for the
     * type of entity class that is being queried.</p>
     *
     * @param <E>         entity class of the attributes that are used as sort criteria.
     * @param entityClass entity class of the attributes that are used as sort criteria.
     * @return a request for the previous page.
     * @throws NoSuchElementException if it is known that there is no previous page.
     *         To avoid this exception, check for a {@code true} result of
     *         {@link #hasPrevious()} before invoking this method.
```

**Review by njr-11 — COMMENTED** [review]
All of the updates look good.  I spotted a soft merge conflict in a few parts of the `Page` JavaDoc after Otavio merged the pull to switch from `null` to `NoSuchElementException`.

**Comment by gavinking** [comment]
@njr-11 committed your suggestions and also proposed some further edits to the javadoc.

**Review by njr-11 — APPROVED** [review]
Everything looks good now. Thanks!

**Comment by njr-11** [comment]
@otaviojava you should review this as well

File: api/src/main/java/jakarta/data/page/Page.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/page/Page.java:36]:
```suggestion
 *                       {@code PageRequest<Vehicle>} pageRequest);
```

File: api/src/main/java/jakarta/data/page/Page.java
Comment (otaviojava) [inline api/src/main/java/jakarta/data/page/Page.java:184]: Pagination with total is a common feature in many NoSQL databases, but it's **not universal** across all NoSQL databases. Whether pagination with total is supported depends on the specific database management system and its features.

I would go with Unsupportedoperationexception when not supported.

**Review by otaviojava — APPROVED** [review]

File: tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/PersistenceEntityTests.java
Comment (otaviojava) [inline tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/PersistenceEntityTests.java:32]: It is breaking the check style, once we are not using it:

```suggestion

````

File: tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/Catalog.java
Comment (otaviojava) [inline tck/src/main/java/ee/jakarta/tck/data/standalone/persistence/Catalog.java:25]: To fix the check style:

```suggestion

```

**Pull request merged**
